wsd/ClientSession.cpp

```diff
@@ -877,8 +877,37 @@ bool ClientSession::_handleInput(const char *buffer, int length)
 
         return true;
     }
-    else if (tokens.equals(0, "dialogevent") ||
-             tokens.equals(0, "formfieldevent") ||
+    else if (tokens.equals(0, "dialogevent"))
+    {
+        if (tokens.size() > 2)
+        {
+            std::string jsonString = tokens.cat("", 2);
+            try
+            {
+                Poco::JSON::Parser parser;
+                const Poco::Dynamic::Var result = parser.parse(jsonString);
+                const auto& object = result.extract<Poco::JSON::Object::Ptr>();
+                const std::string id = object->has("id") ? object->get("id").toString() : "";
+                if (id == "changepass" && _wopiFileInfo && !isDocumentOwner())
+                {
+                    sendTextFrameAndLogError("error: cmd=dialogevent kind=cantchangepass");
+                }
+                else
+                {
+                    return forwardToChild(firstLine, docBroker);
+                }
+            }
+            catch (const std::exception& exception)
+            {
+                return forwardToChild(firstLine, docBroker);
+            }
+        }
+        else
+        {
+            return forwardToChild(firstLine, docBroker);
```

Comment:
Since we always default to `forwardToChild`, probably better to have this as the unconditional fallback at the end, and return early where we don't want to default. The code should be simpler and more readable.